pkg/crdmetrics/collector.go

```diff
@@ -0,0 +1,161 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package crdmetrics
+
+import (
+	"context"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/metric"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/open-telemetry/opentelemetry-operator/apis/v1alpha1"
+	"github.com/open-telemetry/opentelemetry-operator/apis/v1beta1"
+)
+
+type OpenTelemetryCollectorMetrics struct {
+	client       client.Client
+	observations []instancesView
+}
+
+func newOpenTelemetryCollectorMetrics(client client.Client) *OpenTelemetryCollectorMetrics {
+	return &OpenTelemetryCollectorMetrics{
+		client: client,
+	}
+}
+
+func (i *OpenTelemetryCollectorMetrics) Setup() error {
+	meter := otel.Meter(meterName)
+
+	obs, err := newObservation(meter,
+		receivers,
+		"Number of instances using certain receiver",
+		"type",
+		func(instance *collectorInfo) ([]string, bool) {
+			return instance.components.receivers, true
+		})
+	if err != nil {
+		return err
+	}
+	i.observations = append(i.observations, obs)
+
+	obs, err = newObservation(meter,
+		exporters,
+		"Number of instances using certain exporter",
+		"type",
+		func(instance *collectorInfo) ([]string, bool) {
+			return instance.components.exporters, true
+		})
+	if err != nil {
+		return err
+	}
+	i.observations = append(i.observations, obs)
+
+	obs, err = newObservation(meter,
+		processors,
+		"Number of instances using certain processor",
+		"type",
+		func(instance *collectorInfo) ([]string, bool) {
+			return instance.components.processors, true
+
+		})
+	if err != nil {
+		return err
+	}
+	i.observations = append(i.observations, obs)
+
+	obs, err = newObservation(meter,
+		extensions,
+		"Number of instances using certain extension",
+		"type",
+		func(instance *collectorInfo) ([]string, bool) {
+			return instance.components.extensions, true
+
+		})
+	if err != nil {
+		return err
+	}
+	i.observations = append(i.observations, obs)
+
+	obs, err = newObservation(meter,
+		mode,
+		"Collector deployment mode",
+		"mode",
+		func(instance *collectorInfo) ([]string, bool) {
+			return []string{instance.mode}, true
+
+		})
+	if err != nil {
+		return err
+	}
+	i.observations = append(i.observations, obs)
+	instruments := make([]metric.Observable, 0, len(i.observations))
+	for _, o := range i.observations {
+		instruments = append(instruments, o.Gauge)
+	}
+	_, err = meter.RegisterCallback(i.callback, instruments...)
+	return err
+}
+
+func (i *OpenTelemetryCollectorMetrics) callback(ctx context.Context, observer metric.Observer) error {
+	// Reset observations
+	for _, o := range i.observations {
+		o.reset()
+	}
+
+	v1alpha1collectors := &v1alpha1.OpenTelemetryCollectorList{}
+	if err := i.client.List(ctx, v1alpha1collectors); err == nil {
```

Comment:
This can be removed the CRs are stored always in a single version and can be retrieved for both. I would recommend sticking only with v1beta1